File: 스택,큐.힙/이중우선순위큐(PRO).kt
import java.util.*

class Solution {
    fun solution(operations: Array<String>): IntArray {
        var minHeap = PriorityQueue<Int>()
        var maxHeap = PriorityQueue<Int>(Comparator.reverseOrder())
        
        for (op in operations) {
            val (com, num) = op.split(" ")
            
            when(com) {
                "I" -> {
                    minHeap.offer(num.toInt())
                    maxHeap.offer(num.toInt())
                }
                "D" -> {
                    if (num == "1") {
                        val v = maxHeap.poll()
                        minHeap.remove(v)
                    }
                    else {
                        val v = minHeap.poll()
                        maxHeap.remove(v)
                    }
                }
            }
        }
        
        return if (minHeap.size >0) intArrayOf(maxHeap.poll(), minHeap.poll()); else intArrayOf(0, 0)
    }
}

File: 스택,큐.힙/이중우선순위큐(PRO).py
import heapq

def solution(operations):
    answer = []
    
    minHeap = []
    maxHeap = []
    
    for op in operations:
        command, num = op.split()
        
        if command == 'I':
            heapq.heappush(minHeap, int(num))
            heapq.heappush(maxHeap, -int(num))
        else:
            if num == '-1' and minHeap:
                value = heapq.heappop(minHeap)
                maxHeap.remove(-value)
            elif num == '1' and maxHeap:
                value = heapq.heappop(maxHeap)
                minHeap.remove(-value)
    
    return [-heapq.heappop(maxHeap), heapq.heappop(minHeap)] if len(minHeap) > 0 else [0,0]